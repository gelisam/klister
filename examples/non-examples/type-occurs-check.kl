#lang "prelude.kl"

-- accepted, thanks to let-generalization
(example
  (let [xs (nil)]
    (:: xs  -- (the (List A) xs)
        xs  -- (the (List (List A)) xs)
        )))

-- causes the type equality ?1 ~ List ?1, which must be rejected by the
-- occurs-check.
(defun infinite-type (xs)
  (:: xs xs))
