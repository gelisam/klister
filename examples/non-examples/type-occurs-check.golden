Infinite type detected: (MetaPtr 57868) = (List META(MetaPtr 57868))
